### PR TITLE
feat: subscriptions for meta fields 

### PR DESCRIPTION
**Pull request by maan2003**
follow up for #4768 

## Summary by CodeRabbit

- **New Features**
	- Introduced enhanced client metadata management capabilities.
	- Added new functionality for managing and caching metadata fields, with support for continuous updates from various sources.
	- Implemented a `Waiter` functionality for asynchronous task completion tracking.

- **Enhancements**
	- Integrated `reqwest` library for improved HTTP request handling.

- **Documentation**
	- Not explicitly mentioned, but changes suggest updates to how metadata and task management are documented internally.

### CODE REVIEW

File: fedimint-client/src/meta.rs [hunk 1]
```diff
@@ -96,6 +101,20 @@ impl MetaService {
             .find_map(|(_, value)| parse_meta_value_static::<V>(&value).ok())
     }
 
+    pub async fn subscribe_to_updates(&self) -> impl Stream<Item = ()> + '_ {
+        stream! {
+            let mut notify = pin!(self.meta_update_notify.notified());
+            loop {
+                notify.as_mut().await;
+                notify.set(self.meta_update_notify.notified());
```
Comment (maan2003) [inline fedimint-client/src/meta.rs:109]: TIL there is Pin::set

File: fedimint-client/src/meta.rs [hunk 2]
```diff
@@ -113,6 +118,43 @@ impl MetaService {
...
+    /// NOTE: this subscription never ends even after update task is shutdown.
+    /// You should consume this stream in a spawn_cancellable.
+    pub fn subscribe_to_updates(&self) -> impl Stream<Item = ()> + '_ {
+        stream! {
+            let mut notify = pin!(self.meta_update_notify.notified());
+            loop {
+                notify.as_mut().await;
+                notify.set(self.meta_update_notify.notified());
+                // enable waiting for next notification before yield so don't miss
+                // any notifications.
```
Comment (elsirion) [inline fedimint-client/src/meta.rs:130]: Good catch, I was wondering why this looked so complicated.

**Review by elsirion — COMMENTED** [review]
ACK for the last commit, but the meta service has open questions.

File: fedimint-client/src/meta.rs [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+                notify.as_mut().enable();
+                yield ();
+            }
+        }
+    }
+
+    /// NOTE: this subscription never ends even after update task is shutdown.
+    /// You should consume this stream in a spawn_cancellable.
+    pub async fn subscribe_to_field<'a, V: DeserializeOwned + 'static>(
```
Comment (maan2003) [inline fedimint-client/src/meta.rs:145]: to make sure that 
```
let stream = meta.subscribe_to_field().await;
let initial = meta.get_field("foo").await;

// .. use stream
```
doesn't miss any updates

**Review by elsirion — DISMISSED** [review]

File: fedimint-client/src/meta.rs (hunk 3 above)
Comment (maan2003) [reply fedimint-client/src/meta.rs:145]: fixed, just changed `subscribe_to_field` slightly, see docs

File: fedimint-client/src/meta.rs [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        };
+        Ok(*ret)
+    } else {
+        res
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+pub trait MetaSource: MaybeSend + MaybeSync {
+    /// This must be stable.
```
Comment (dpc) [inline fedimint-client/src/meta.rs:53]: Wouldn't it be easier to give use `&'stat str`s for this?

Comment (maan2003) [reply fedimint-client/src/meta.rs:53]: agreed, done

**Comment by maan2003** [comment]
rebased on parent

**Review by elsirion — DISMISSED** [review]

**Review by dpc — APPROVED** [review]

**Review by elsirion — APPROVED** [review]

**Comment by fedimint-backports** [comment]
Backport failed for `releases/v0.3`, because it was unable to cherry-pick the commit(s).

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin releases/v0.3
git worktree add -d .worktree/backport-4787-to-releases/v0.3 origin/releases/v0.3
cd .worktree/backport-4787-to-releases/v0.3
git switch --create backport-4787-to-releases/v0.3
git cherry-pick -x edf977d04a3bf8925fa41fd7f62909fcdd7abde1
```

**Pull request merged**